koku/subs/subs_data_messenger.py

```diff
@@ -34,6 +43,36 @@ def __init__(self, context, schema_name, tracing_id):
         self.account_id = subs_cust.account_id
         self.org_id = subs_cust.org_id
         self.download_path = mkdtemp(prefix="subs")
+        self.instance_map = {}
+
+    def determine_azure_instance_id(self, row):
+        """For Azure we have to query the instance id if it is not provided via tag."""
+        if row["resourceid"] in self.instance_map:
+            return self.instance_map.get(row["resourceid"])
+        # this column comes from a user defined tag allowing us to avoid querying Azure if its present.
+        if row["subs_instance"] != "":
+            instance_id = row["subs_instance"]
+        # attempt to query azure for instance id
+        else:
+            # if its a local Azure provider, don't query Azure
+            if self.local_prov:
+                return ""
+            prov = Provider.objects.get(uuid=row["source"])
+            credentials = prov.account.get("credentials")
+            subscription_id = credentials.get("subscription_id")
+            tenant_id = credentials.get("tenant_id")
+            client_id = credentials.get("client_id")
+            client_secret = credentials.get("client_secret")
+            _factory = AzureClientFactory(subscription_id, tenant_id, client_id, client_secret)
+            compute_client = _factory.compute_client
+            response = compute_client.virtual_machines.get(
```

Comment:
I'm sure you already have this covered. But for my curiosity, does this client query need additional permissions exposed on the customer side? I assume it does since we only have something like storage blob and cost management reader roles right now.

Reply:
Yeah, there's an additional `Reader` role that will need to be added. It will be called out in docs